torch_np/_funcs.py

```diff
@@ -503,3 +1698,203 @@ def median(
+
+    result = torch.diff(a, n, axis=axis, prepend=prepend, append=append)
+
+    return result
+
+
+# ### math functions ###
+
+
+@normalizer
+def angle(z: ArrayLike, deg=False):
+    result = torch.angle(z)
+    if deg:
+        result = result * 180 / torch.pi
+    return result
+
+
+@normalizer
+def sinc(x: ArrayLike):
+    result = torch.sinc(x)
+    return result
+
+
+@normalizer
+def real(a: ArrayLike):
+    result = torch.real(a)
+    return result
+
+
+@normalizer
+def imag(a: ArrayLike):
+    if a.is_complex():
+        result = a.imag
+    else:
+        result = torch.zeros_like(a)
+    return result
+
+
+@normalizer
+def round_(a: ArrayLike, decimals=0, out: Optional[NDArray] = None) -> OutArray:
+    if a.is_floating_point():
+        result = torch.round(a, decimals=decimals)
+    elif a.is_complex():
+        # RuntimeError: "round_cpu" not implemented for 'ComplexFloat'
+        result = (
+            torch.round(a.real, decimals=decimals)
+            + torch.round(a.imag, decimals=decimals) * 1j
+        )
+    else:
+        # RuntimeError: "round_cpu" not implemented for 'int'
+        result = a
+    return result, out
+
+
+around = round_
+round = round_
+
+
+@normalizer
+def real_if_close(a: ArrayLike, tol=100):
+    # XXX: copies vs views; numpy seems to return a copy?
+    if not torch.is_complex(a):
+        return a
+    if tol > 1:
+        # Undocumented in numpy: if tol < 1, it's an absolute tolerance!
+        # Otherwise, tol > 1 is relative tolerance, in units of the dtype epsilon
+        # https://github.com/numpy/numpy/blob/v1.24.0/numpy/lib/type_check.py#L577
+        tol = tol * torch.finfo(a.dtype).eps
+
+    mask = torch.abs(a.imag) < tol
+    return a.real if mask.all() else a
+
+
+@normalizer
+def iscomplex(x: ArrayLike):
+    if torch.is_complex(x):
+        return torch.as_tensor(x).imag != 0
+    result = torch.zeros_like(x, dtype=torch.bool)
+    if result.ndim == 0:
+        result = result.item()
+    return result
+
+
+@normalizer
+def isreal(x: ArrayLike):
+    if torch.is_complex(x):
+        return torch.as_tensor(x).imag == 0
+    result = torch.ones_like(x, dtype=torch.bool)
+    if result.ndim == 0:
+        result = result.item()
+    return result
+
+
+@normalizer
+def iscomplexobj(x: ArrayLike):
+    result = torch.is_complex(x)
+    return result
+
+
+@normalizer
+def isrealobj(x: ArrayLike):
+    result = not torch.is_complex(x)
+    return result
+
+
+@normalizer
+def isneginf(x: ArrayLike, out: Optional[NDArray] = None):
+    result = torch.isneginf(x, out=out)
+    return result
+
+
+@normalizer
+def isposinf(x: ArrayLike, out: Optional[NDArray] = None):
+    result = torch.isposinf(x, out=out)
+    return result
+
+
+@normalizer
+def i0(x: ArrayLike):
+    result = torch.special.i0(x)
+    return result
+
+
+@normalizer(return_on_failure=False)
+def isscalar(a: ArrayLike):
+    # XXX: this is a stub
+    if a is False:
+        return a
```

Comment:
What does this construction mean?

Reply:
`if a is False: return False`?

Reply:
Yeah, but `a` is an `ArrrayLike`, when is it going to be the literal `False`?

Reply:
OK, fair enough. Thanks for spotting this. Removed.

Reply:
This uncovered a latent bug:
```
    @normalizer(return_on_failure=False)
    def isscalar(a: ArrayLike):
        # XXX: this is a stub
>       return a.numel() == 1
E       AttributeError: 'bool' object has no attribute 'numel'
```
It may be better to just implement this function without the normalizer and remove the `return_on_failure` flag, as you initially had?

Reply:
All right. Here's what's going on: `isscalar(np.int)`  (`np.int`is not a scalar!) fails to convert its argument to tensor, and the `return_on_failure` code path feeds `False` into the the body of the function. This is in fact requested by the normalizer argument three lines above, via `normalizer(return_on_failure=False)`.  So, can either 
- put back the `if a is False: return False`  clause + a code comment to not confuse future readers
- rip the `return_on_failure` code path and normalize manually (and possibly will need to normalize manually other functions from the not-yet-implemented numpy apis, who knows)

Your preference @lezcano ?

Reply:
Probably the simplest option is the second one, as this function can be implemented as
```python
def isscalar(a):
    try:
        return asarray(a).size == 1
    except:
        return False
```
and it's very much one of a kind.